File: backend/app/api/bracket_orders.py
from fastapi import APIRouter, HTTPException, Depends, status
from fastapi.security import HTTPBearer, HTTPAuthorizationCredentials
from typing import List, Optional

from ..models.bracket_order import (
    BracketOrderCreate,
    BracketOrderResponse,
    BracketOrderUpdate,
    BracketOrderValidationError
)
from ..services.bracket_order_service import bracket_order_service

router = APIRouter()
security = HTTPBearer()

@router.post("/", response_model=BracketOrderResponse)
async def create_bracket_order(
    order: BracketOrderCreate,
    # credentials: HTTPAuthorizationCredentials = Depends(security)  # Temporarily disabled for testing
):
    """Create a new bracket order"""
    try:
        bracket_order = bracket_order_service.create_bracket_order(order)
        return bracket_order
    except BracketOrderValidationError as e:
        raise HTTPException(
            status_code=status.HTTP_400_BAD_REQUEST,
            detail=str(e)
        )
    except Exception as e:
        raise HTTPException(
            status_code=status.HTTP_500_INTERNAL_SERVER_ERROR,
            detail=f"Failed to create bracket order: {str(e)}"
        )

@router.get("/", response_model=List[BracketOrderResponse])
async def get_bracket_orders(
    symbol: Optional[str] = None,
    credentials: HTTPAuthorizationCredentials = Depends(security)
):
    """Get all bracket orders, optionally filtered by symbol"""
    try:
        orders = bracket_order_service.get_bracket_orders(symbol=symbol)
        return orders
    except Exception as e:
        raise HTTPException(
            status_code=status.HTTP_500_INTERNAL_SERVER_ERROR,
            detail=f"Failed to retrieve bracket orders: {str(e)}"
        )

@router.get("/{order_id}", response_model=BracketOrderResponse)
async def get_bracket_order(
    order_id: str,
    credentials: HTTPAuthorizationCredentials = Depends(security)
):
    """Get a specific bracket order by ID"""
    order = bracket_order_service.get_bracket_order(order_id)
    if not order:
        raise HTTPException(
            status_code=status.HTTP_404_NOT_FOUND,
            detail="Bracket order not found"
        )
    return order

@router.put("/{order_id}", response_model=BracketOrderResponse)
async def update_bracket_order(
    order_id: str,
    updates: BracketOrderUpdate,
    credentials: HTTPAuthorizationCredentials = Depends(security)
):
    """Update a bracket order (only for pending orders)"""
    try:
        updated_order = bracket_order_service.update_bracket_order(order_id, updates)
        if not updated_order:
            raise HTTPException(
                status_code=status.HTTP_404_NOT_FOUND,
                detail="Bracket order not found or cannot be updated"
            )
        return updated_order
    except BracketOrderValidationError as e:
        raise HTTPException(
            status_code=status.HTTP_400_BAD_REQUEST,
            detail=str(e)
        )
    except Exception as e:
        raise HTTPException(
            status_code=status.HTTP_500_INTERNAL_SERVER_ERROR,
            detail=f"Failed to update bracket order: {str(e)}"
        )

@router.delete("/{order_id}")
async def cancel_bracket_order(
    order_id: str,
    credentials: HTTPAuthorizationCredentials = Depends(security)
):
    """Cancel a bracket order"""
    try:
        success = bracket_order_service.cancel_bracket_order(order_id)
        if not success:
            raise HTTPException(
                status_code=status.HTTP_404_NOT_FOUND,
                detail="Bracket order not found or cannot be cancelled"
            )
        return {"message": f"Bracket order {order_id} cancelled successfully"}
    except Exception as e:
        raise HTTPException(
            status_code=status.HTTP_500_INTERNAL_SERVER_ERROR,
            detail=f"Failed to cancel bracket order: {str(e)}"
        )

@router.get("/{symbol}/market-price")
async def get_market_price(
    symbol: str,
    credentials: HTTPAuthorizationCredentials = Depends(security)
):
    """Get current market price for a symbol"""
    try:
        price = bracket_order_service.get_current_market_price(symbol)
        return {"symbol": symbol, "price": str(price)}
    except Exception as e:
        raise HTTPException(
            status_code=status.HTTP_500_INTERNAL_SERVER_ERROR,
            detail=f"Failed to get market price: {str(e)}"
        )

File: frontend/src/services/bracketOrders.ts
import api from './api'
import { BracketOrderCreate, BracketOrderResponse } from '../types/bracketOrder'
import { getMockOrders, addMockOrder, removeMockOrder, updateMockOrder } from './mockBracketOrders'

// Flag to track if backend is available
let backendAvailable = true

export const bracketOrdersApi = {
  // Create a new bracket order
  create: async (order: BracketOrderCreate): Promise<BracketOrderResponse> => {
    try {
      const response = await api.post('/bracket-orders', order)
      backendAvailable = true
      return response.data
    } catch (error: any) {
      if (error?.response?.status === 404 || error?.code === 'ERR_NETWORK') {
        // Backend not available, use mock
        backendAvailable = false
        const mockOrder: BracketOrderResponse = {
          ...order,
          id: `mock-${Date.now()}`,
          status: 'pending',
          created_at: new Date().toISOString(),
          updated_at: new Date().toISOString()
        }
        addMockOrder(mockOrder)
        return mockOrder
      }
      throw error
    }
  },

  // Get all bracket orders
  getAll: async (symbol?: string): Promise<BracketOrderResponse[]> => {
    try {
      const params = symbol ? { symbol } : {}
      const response = await api.get('/bracket-orders', { params })
      backendAvailable = true
      return response.data
    } catch (error: any) {
      if (error?.response?.status === 404 || error?.code === 'ERR_NETWORK') {
        // Backend not available, return empty array or mock orders
        backendAvailable = false
        return getMockOrders().filter(o => !symbol || o.symbol === symbol)
      }
      throw error
    }
  },

  // Get a specific bracket order
  getById: async (orderId: string): Promise<BracketOrderResponse> => {
    const response = await api.get(`/bracket-orders/${orderId}`)
    return response.data
  },

  // Update a bracket order
  update: async (orderId: string, updates: any): Promise<BracketOrderResponse> => {
    try {
      const response = await api.put(`/bracket-orders/${orderId}`, updates)
      backendAvailable = true
      return response.data
    } catch (error: any) {
      if ((error?.response?.status === 404 || error?.code === 'ERR_NETWORK') && orderId.startsWith('mock-')) {
        // Handle mock order update
        backendAvailable = false
        const mockOrder = getMockOrders().find(o => o.id === orderId)
        if (mockOrder) {
          const updated = { ...mockOrder, ...updates, updated_at: new Date().toISOString() }
          updateMockOrder(orderId, updated)
          return updated
        }
      }
      throw error
    }
  },

  // Cancel a bracket order
  cancel: async (orderId: string): Promise<{ message: string }> => {
    try {
      const response = await api.delete(`/bracket-orders/${orderId}`)
      backendAvailable = true
      return response.data
    } catch (error: any) {
      if (error?.response?.status === 404 || (error?.code === 'ERR_NETWORK' && orderId.startsWith('mock-'))) {
        // If it's a mock order or backend not available, just remove from mock store
        backendAvailable = false
        removeMockOrder(orderId)
        return { message: 'Order cancelled successfully' }
      }
      throw error
    }
  },

  // Get current market price
  getMarketPrice: async (symbol: string): Promise<{ symbol: string; price: string }> => {
    const response = await api.get(`/bracket-orders/${symbol}/market-price`)
    return response.data
  }
}